File: trunk/server/restserver.php
#!/php -q
<?php
/*  > php -q server.php  */

//-----------------------------------------------
// Configurations
//-----------------------------------------------

error_reporting(E_ALL);
set_time_limit(0);
ob_implicit_flush();

$master = WebSocket("0.0.0.0", 8080);
$sockets = array($master);
$users = array();
$debug = true;

$db_hostname = '127.0.0.1';
$db_database = 'xxxxx';
$db_username = 'root';
$db_password = 'somepass';

//-----------------------------------------------
// Main: connect database, start websocket server
//-----------------------------------------------

$db_server = connect_db();
if (!$db_server) {
    say("Failed to connect to database");
    exit();
}

while (true) {
    $changed = $sockets;
    socket_select($changed, $write = NULL, $except = NULL, NULL); //have some issues here for connect
    foreach ($changed as $socket) {
        if ($socket == $master) {
            $client = socket_accept($master);
            if ($client < 0) {
                console("socket_accept() failed");
                continue;
            } else {
                connect($client);
            }
        } else {
            $bytes = socket_recv($socket, $buffer, 2048, 0);
            if ($bytes == 0) {
                on_disconnect(getuserbysocket($socket));
                disconnect($socket);
            } else {
                $user = getuserbysocket($socket);
                if (!$user->handshake) {
                    dohandshake($user, $buffer);
                } else {
                    process($user, $buffer);
                }
            }
        }
    }
}

//-----------------------------------------------
// MySQL database connector
//-----------------------------------------------

function connect_db() {
    global $db_hostname, $db_username, $db_password, $db_database;
    $db_server = mysql_connect($db_hostname, $db_username, $db_password);
    if (!$db_server) {
        say("ERROR: unable to connect to MySQL: " . mysql_error());
        return NULL;
    }
    mysql_select_db($db_database, $db_server);
    
    // cleanup the subscribe table, since there are no subscription on startup
    mysql_query("DELETE FROM subscribe");
    mysql_query("DELETE FROM resource WHERE cid != ''");
    
    return $db_server;
}

function disconnect_db() {
    mysql_close($db_server);
}

//-----------------------------------------------
// Our websocket data processing
//-----------------------------------------------

function process($user, $msg) {
    $action = unwrap($msg);
    say("< " . $action);

    $request = json_decode($action, true);

    if (empty($request)) {
        say("ERROR: invalid request body");
        return;
    }

    if (!array_key_exists("method", $request)
        || !array_key_exists("resource", $request)
        || !array_key_exists("msg_id", $request)) {
        say("ERROR: missing mandatory property");
        return;
    }

    $method = $request["method"];
    $response = NULL;

    if ($method == "POST" && $request["resource"] == "/slideshare") {
        $response = get_slideshare($request);  
    } else if ($method == "POST") {
        $response = do_post($user, $request);
    } else if ($method == "PUT") {
        $response = do_put($user, $request);
    } else if ($method == "GET") {
        $response = do_get($user, $request);
    } else if ($method == "DELETE") {
        $response = do_delete($user, $request);
    } else if ($method == "SUBSCRIBE") {
        $response = do_subscribe($user, $request);
    } else if ($method == "NOTIFY") {
        $response = do_notify($user, $request);
    } else {
        // this is an unknown request
        $response = array("code" => "failed", "reason" => "unknown command " . $method . " " . $resource);
    }

    $response['msg_id'] = $request['msg_id'];
    header("Content-type: application/json");
    send($user->socket, json_encode($response));
}


function on_disconnect($user) {
    $cid = mysql_real_escape_string($user->id);
    
    // first delete all subscriptions of this client
    mysql_command(sprintf("DELETE FROM subscribe WHERE cid='%s'", $cid));
    
    // get all the transient resources of this client
    $result = mysql_query(sprintf("SELECT rid FROM resource WHERE cid='%s'", $cid));
    $result_count = mysql_numrows($result);
    $transient_resources = array();
    for ($j=0; $j < $result_count; ++$j) {
        $row = mysql_fetch_row($result);
        array_push($transient_resources, $row[0]);
    }
    //mysql_freeresult($result);
    
    // remove all the transient resources of this client
    mysql_command(sprintf("DELETE FROM resource WHERE cid='%s'", $cid));
    
    // notify all the subscribers of those transient resources
    foreach ($transient_resources as $rid) {
        do_notify($user, $rid, "DELETE");
    }
    return array('code' =>'success');
}

//-----------------------------------------------
// Generic RESTful web service
//-----------------------------------------------

function do_post($user, $request) {
    // TODO: validate that resource is in correct format.
    $parent = $request["resource"];
    $id = array_key_exists("id", $request) ? $request["id"] : uniqid();
    $resource = $parent . "/" . $id;
    $type = array_key_exists("type", $request) ? $request["type"] : "application/json";
    $entity = json_encode($request["entity"]);
    $cid = array_key_exists("persistent", $request) && $request["persistent"] ? '' : $user->id;
    
    $result = mysql_query(sprintf("INSERT INTO resource (rid, prid, type, entity, cid) VALUES ('%s', '%s', '%s', '%s', '%s')",
        mysql_real_escape_string($resource), mysql_real_escape_string($parent),
        mysql_real_escape_string($type), mysql_real_escape_string($entity),
        mysql_real_escape_string($cid)));
    if (!$result) {
        return array('code' => 'failed', 'reason' => 'failed to insert this resource');
    }
    //mysql_freeresult($result);
    
    do_notify($user, $resource, "POST");
    return array('code' => 'success', 'id' => $id);
}

function do_put($user, $request) {
    $resource = $request["resource"];
    $parent = get_parent($resource);
    $type = array_key_exists("type", $request) ? $request["type"] : "application/json";
    $entity = json_encode($request["entity"]);
    $cid = array_key_exists("persistent", $request) && $request["persistent"] ? '' : $user->id;
    
    $result = mysql_query(sprintf("REPLACE INTO resource (rid, prid, type, entity, cid) VALUES ('%s', '%s', '%s', '%s', '%s')",
        mysql_real_escape_string($resource), mysql_real_escape_string($parent),
        mysql_real_escape_string($type), mysql_real_escape_string($entity),
        mysql_real_escape_string($cid)));
    if (!$result) {
        return array('code' => 'failed', 'reason' => 'failed to replace this resource');
    }
    //mysql_freeresult($result);
    
    do_notify($user, $resource, "PUT");
    return array('code' => 'success');
}


function do_get($user, $request) {
    $resource = $request['resource'];
    $result = mysql_query(sprintf("SELECT type, entity FROM resource WHERE rid='%s'",
                        mysql_real_escape_string($resource)));
    if (!$result) {
        return array('code' => 'failed', 'reason' => 'failed to get this resource');
    }
    
    $result_count = mysql_num_rows($result);
    if ($result_count) {
        // found this resource with valid value. There should be only one value.
        $row = mysql_fetch_row($result);
        $type = $row[0];
        $entity = json_decode($row[1]);
        //mysql_freeresult($result);
        
        return array('code' => 'success', 'resource' => $resource,
                     'type' =>  $type, 'entity' => $entity);
    }
    
    $result = mysql_query(sprintf("SELECT rid FROM resource WHERE prid='%s'",
                        mysql_real_escape_string($resource)));
    if (!$result) {
        return array('code' => 'failed', 'reason' => 'failed to get child resources');
    }
    
    $response = array();
    $result_count = mysql_num_rows($result);
    for ($j=0; $j<$result_count; ++$j) {
        $row = mysql_fetch_row($result);
        if (substr($row[0], 0, strlen($resource)) === $resource)
            array_push($response, substr($row[0], strlen($resource) + 1));
        else
            array_push($response, $row[0]);
    }
    //mysql_freeresult($result);
    
    if (count($response) > 0) {
        return array('code' => 'success', 'resource' => $resource,
                     'type' => 'application/json', 'entity' => $response);
    }
    
    // try to get sub resources
    return array('code' => 'failed', 'reason' => 'no value found for this resource');
}

function do_delete($user, $request) {
    $resource = $request['resource'];
    
    // disable delete if the child resource exists
    $result = mysql_query(sprintf("SELECT rid FROM resource WHERE prid='%s'",
                        mysql_real_escape_string($resource)));
    $result_count = mysql_num_rows($result);
    //mysql_freeresult($result);
    
    if ($result_count) {
        return array('code' => 'failed', 'reason' => 'this parent resource has children');
    }
    
    // remove the specific resource.
    mysql_command(sprintf("DELETE FROM resource WHERE rid='%s'",
        mysql_real_escape_string($resource)));
    
    do_notify($user, $resource, "DELETE");
    return array('code' => 'success');
}

function do_subscribe($user, $request) {
    $cid = mysql_real_escape_string($user->id);
    $resource = $request['resource'];
    
    $result = mysql_command(sprintf("REPLACE INTO subscribe (rid, cid) VALUES ('%s', '%s')",
                mysql_real_escape_string($resource), $cid));
    if (!$result) {
        return array('code' => 'failed', 'reason' => 'failed to subscribe the client to the resource');
    }
    
    return array('code' => 'success');
}

function do_notify($user, $request, $method = NULL) {
    $notify = array("from" => $user->id);
    
    if ($method != NULL) {
        // notification due to POST, PUT or DELETE
        $resource = $request;
        $notify = array("notify" => $method, "resource" => $resource, "type" => NULL, "entity" => NULL);

        if ($method == "PUT" || $method == "POST") {
            $result = mysql_query(sprintf("SELECT type, entity FROM resource WHERE rid='%s'",
                mysql_real_escape_string($resource)));
            if (!$result) {
                say("failed to get this resource");
                return array('code' => 'failed', 'reason' => 'failed to get this resource');
            }
            $result_count = mysql_numrows($result);
            if ($result_count) {
                $row = mysql_fetch_row($result);
                $notify["type"] = $row[0];
                $notify["entity"] = json_decode($row[1]);
            }
            //mysql_freeresult($result);
        }
        // TODO: also send to parent resource
    } else {
        // end to end notify from one client to others
        $notify = array("notify" => "NOTIFY", "resource" => $request["resource"], "data" => $request["data"]);
    }
    
    $param = json_encode($notify);
    $param = str_replace("\/", "/", $param);

    $result = mysql_query(sprintf("SELECT cid FROM subscribe WHERE rid='%s'",
                        mysql_real_escape_string($notify["resource"])));
    if (!$result) {
        say("failed to get this resource subscribers");
        return array('code' => 'failed', 'reason' => 'failed to get this resource subscribers');
    }
    $result_count = mysql_numrows($result);
    
    $sent_count = 0;
    for ($j=0; $j<$result_count; ++$j) {
        $row = mysql_fetch_row($result);
        $target = getuserbyid($row[0]);
        if ($target == null) {
            say("invalid user for " . $row[0]);
        } else {
            send($target->socket, $param);
            ++$sent_count;
        }
    }
    //mysql_freeresult($result);
    
    if ($method == "POST" || $method == "PUT" || $method == "DELETE") {
        $parent = get_parent($notify["resource"]);
        $change = array("notify" => "UPDATE", "resource" => $parent, "type" => $notify["type"], "entity" => $notify["entity"]);
        $child = $notify["resource"];
        $index = strrpos($child, "/");
        if ($index)
            $child = substr($child, $index+1);
        if ($method == "POST")
            $change["create"] = $child;
        else if ($method == "PUT")
            $change["update"] = $child;
        else if ($method == "DELETE")
            $change["delete"] = $child;
        $result = mysql_query(sprintf("SELECT cid FROM subscribe WHERE rid='%s'",
                                      mysql_real_escape_string($parent)));
        $result_count = mysql_numrows($result);
        $param = json_encode($change);
        // JSON specified escapting / but we don't accept that in Javascript
        $param = str_replace("\/", "/", $param);
        
        say("change=" . $change . " param=" . $param);
        for ($k=0; $k<$result_count; ++$k) {
            $row = mysql_fetch_row($result);
            $target = getuserbyid($row[0]);
            if ($target == null) {
                say("invalid user for " . $row[0]);
            } else {
                send($target->socket, $param);
                ++$sent_count;
            }
            ++$sent_count;
        }
        //mysql_freeresult($result);
    }
    
    if ($sent_count == 0) {
        say('notify could not send to anyone');
        return array('code' => 'failed', 'reason' => 'no available user to send notification to');
    }
    
    say('notify sent to ' . count($sent_count) . ' items');
    return array('code' => 'success', 'sent_count' => $sent_count);
}

function get_parent($resource) {
    $index = strrpos($resource, "/");
    if ($index == FALSE)
        return '';
    return substr($resource, 0, $index);
}

function mysql_command($query) {
    $result = mysql_query($query);
    if ($result) {
        //mysql_freeresult($result);
        return TRUE;
    } else {
        return FALSE;
    }
}

//-----------------------------------------------
// RPC commands
//-----------------------------------------------

// this PHP function borrowed from
// http://hasin.wordpress.com/2008/02/09/hacking-slidesharenet-using-php/

function get_slideshare($request) {
    $url = $request["url"];
    $page = file_get_contents($url);
    $pattern = "~doc=([\w-]+)~";
    preg_match($pattern, $page, $matches);
    $xmlurl = "http://s3.amazonaws.com/slideshare/{$matches[1]}.xml";
    $sxml = simplexml_load_file($xmlurl);
    $result = array();
    foreach ($sxml->Slide as $slide)
        array_push($result, (string) $slide['Src']);
    say("result=" . $result);
    return array('code' => 'success', 'entity' => $result);
}

//-----------------------------------------------
// Third-party web socket related code
//-----------------------------------------------

function WebSocket($address, $port) {
    $master = socket_create(AF_INET, SOCK_STREAM, SOL_TCP) or die("socket_create() failed");
    socket_set_option($master, SOL_SOCKET, SO_REUSEADDR, 1) or die("socket_option() failed");
    socket_bind($master, $address, $port) or die("socket_bind() failed");
    socket_listen($master, 20) or die("socket_listen() failed");
    echo "Server Started : " . date('Y-m-d H:i:s') . "\n";
    echo "Master socket  : " . $master . "\n";
    echo "Listening on   : " . $address . " port " . $port . "\n\n";
    return $master;
}

function connect($socket) {
    global $sockets, $users;
    $user = new User();
    $user->id = uniqid();
    $user->socket = $socket;
    array_push($users, $user);
    array_push($sockets, $socket);
    console($socket . " CONNECTED!");
}

function disconnect($socket) {
    global $sockets, $users;
    $found = null;
    $n = count($users);
    for ($i = 0; $i < $n; $i++) {
        if ($users[$i]->socket == $socket) {
            $found = $i;
            break;
        }
    }
    if (!is_null($found)) {
        array_splice($users, $found, 1);
    }
    $index = array_search($socket, $sockets);
    socket_close($socket);
    console($socket . " DISCONNECTED!");
    if ($index >= 0) {
        array_splice($sockets, $index, 1);
    }
}

function dohandshake($user, $buffer) {
    console("\nRequesting handshake...");
    console($buffer);
    list($resource, $host, $origin, $strkey1, $strkey2, $data) = getheaders($buffer);
    console("Handshaking...");

    $pattern = '/[^\d]*/';
    $replacement = '';
    $numkey1 = preg_replace($pattern, $replacement, $strkey1);
    $numkey2 = preg_replace($pattern, $replacement, $strkey2);

    $pattern = '/[^ ]*/';
    $replacement = '';
    $spaces1 = strlen(preg_replace($pattern, $replacement, $strkey1));
    $spaces2 = strlen(preg_replace($pattern, $replacement, $strkey2));

    if ($spaces1 == 0 || $spaces2 == 0 || $numkey1 % $spaces1 != 0 || $numkey2 % $spaces2 != 0) {
        socket_close($user->socket);
        console('failed');
        return false;
    }

    $ctx = hash_init('md5');
    hash_update($ctx, pack("N", $numkey1 / $spaces1));
    hash_update($ctx, pack("N", $numkey2 / $spaces2));
    hash_update($ctx, $data);
    $hash_data = hash_final($ctx, true);

    $upgrade = "HTTP/1.1 101 WebSocket Protocol Handshake\r\n" .
            "Upgrade: WebSocket\r\n" .
            "Connection: Upgrade\r\n" .
            "Sec-WebSocket-Origin: " . $origin . "\r\n" .
            "Sec-WebSocket-Location: ws://" . $host . $resource . "\r\n" .
            "\r\n" .
            $hash_data;

    socket_write($user->socket, $upgrade . chr(0), strlen($upgrade . chr(0)));
    $user->handshake = true;
    console($upgrade);
    console("Done handshaking...");
    return true;
}

function getheaders($req) {
    $r = $h = $o = null;
    if (preg_match("/GET (.*) HTTP/", $req, $match)) {
        $r = $match[1];
    }
    if (preg_match("/Host: (.*)\r\n/", $req, $match)) {
        $h = $match[1];
    }
    if (preg_match("/Origin: (.*)\r\n/", $req, $match)) {
        $o = $match[1];
    }
    if (preg_match("/Sec-WebSocket-Key2: (.*)\r\n/", $req, $match)) {
        $key2 = $match[1];
    }
    if (preg_match("/Sec-WebSocket-Key1: (.*)\r\n/", $req, $match)) {
        $key1 = $match[1];
    }
    if (preg_match("/\r\n(.*?)\$/", $req, $match)) {
        $data = $match[1];
    }
    return array($r, $h, $o, $key1, $key2, $data);
}

function getuserbysocket($socket) {
    global $users;
    $found = null;
    foreach ($users as $user) {
        if ($user->socket == $socket) {
            $found = $user;
            break;
        }
    }
    return $found;
}

function getuserbyid($wsid) {
    global $users;
    $found = null;
    foreach ($users as $user) {
        if ($user->id == $wsid) {
            $found = $user;
            break;
        }
    }
    return $found;
}

function send($client, $msg) {
    say(">".$msg);
    $msg = wrap($msg);
    socket_write($client, $msg, strlen($msg));
}

function say($msg="") {
    echo $msg . "\n";
}

function wrap($msg="") {
   return chr(0) . $msg . chr(255);
  //  return $msg;
}

function unwrap($msg="") {
    return substr($msg, 1, strlen($msg) - 2);
}

function console($msg="") {
    global $debug;
    if ($debug) {
        echo $msg . "\n";
    }
}

class User {
    var $id;
    var $socket;
    var $handshake;
}
?>

File: trunk/server/restserver.py
# This is the Python restserver module used in the websocket server by
# restserver_wsh.py and AJAX server by ajaxserver.py.

import logging, sqlite3, json, time, random, re

logger = logging.getLogger('restserver')

class Database():
    def __init__(self, filename="restserver.db"):
        self.conn = sqlite3.connect(filename, check_same_thread=False)
        self.cursor = self.conn.cursor()
        self._create()
        
    def _create(self):
        try:
            self.commit('''CREATE TABLE resource (
                rid varchar(1024) PRIMARY KEY NOT NULL DEFAULT '',
                prid varchar(1024) NOT NULL DEFAULT '',
                type varchar(64) NOT NULL DEFAULT 'application/json',
                entity blob,
                cid varchar(25)
            )''')
            self.commit('''create table subscribe (
                rid varchar(1024) NOT NULL DEFAULT '',
                cid varchar(25) NOT NULL DEFAULT '',
                PRIMARY KEY (rid, cid)
            )''')
            logger.debug('Database created')
        except sqlite3.OperationalError:
            logger.debug('Database already created')
        

    def reset(self):
        # cleanup the subscribe table, since there are no subscription on startup
        self.commit("DELETE FROM subscribe");
        self.commit("DELETE FROM resource WHERE cid != ''");
        
    def close(self):
        if self.cursor:
            self.cursor.close()
            self.cursor = None
            
    def commit(self, *args):
        logger.debug('commit%r', args)
        self.cursor.execute(*args)
        self.conn.commit()
        
    def iterate(self, *args):
        logger.debug('iterate%r', args)
        return self.cursor.execute(*args)
        
    def fetchone(self, *args):
        logger.debug('fetchone%r', args)
        self.cursor.execute(*args)
        result = self.cursor.fetchone()
        logger.debug('fetchone%r=>\n  %r', args, result)
        return result
        
    def fetchall(self, *args):
        logger.debug('fetchall%r', args)
        self.cursor.execute(*args)
        result = self.cursor.fetchall()
        logger.debug('fetchall%r=>\n  %s', args, '\n  '.join(['%r'%(x,) for x in result]))
        return result

def uniqid():
    return str(int(time.time()) * 1000 + random.randint(0, 999))


class Handler():
    def __init__(self, db):
        self.db = db

    def POST(self, request):
        parent, ctype, entity, persistent = request['resource'], request.get('type', 'application/json'), \
            json.dumps(request.get('entity', {})), request.get('persistent', False)
        rid = request['id'] if 'id' in request else uniqid()
        resource = parent + '/' + rid
        cid = '' if persistent else self.id
        try:
            self.db.commit('INSERT INTO resource (rid, prid, type, entity, cid) VALUES (?, ?, ?, ?, ?)',
                (resource, parent, ctype, entity, cid))
        except:
            logger.exception('failed to insert resource')
            return dict(code='failed', reason='failed to insert this resource')
        self.NOTIFY(resource, 'POST')
        return dict(code='success', id=rid)
    
    def PUT(self, request):
        resource, attr, ignore = self._parse(request['resource'])
        ctype, entity, persistent = request.get('type', 'application/json'), \
            json.dumps(request.get('entity', {})), request.get('persistent', False)
        if attr:
            result = None
            try:
                result = self.db.fetchone('SELECT type, entity FROM resource WHERE rid=?', (resource,))
            except:
                logger.exception('failed to get resource')
            if not result:
                return dict(code='failed', reason='failed to get the resource')
            
            result = json.loads(result[1])
            result[attr] = request.get('entity', None);
            entity = json.dumps(result)
            try:
                self.db.commit('UPDATE resource SET entity=? WHERE rid=?', (entity, resource))
            except:
                logger.exception('failed to replace resource attribute')
                return dict(code='failed', reason='failed to replace resource attribute')
        else:
            parent = self.get_parent(resource)
            cid = '' if persistent else self.id
            try:
                self.db.commit('REPLACE INTO resource (rid, prid, type, entity, cid) VALUES (?, ?, ?, ?, ?)',
                    (resource, parent, ctype, entity, cid))
            except:
                logger.exception('failed to replace resource')
                return dict(code='failed', reason='failed to replace this resource')
        self.NOTIFY(resource, 'PUT')
        return dict(code='success')
    
    def GET(self, request):
        resource, attr, params = self._parse(request['resource'])
        if attr:
            result = None
            try:
                result = self.db.fetchone('SELECT type, entity FROM resource WHERE rid=?', (resource,))
                entity = json.loads(result[1])
                if attr in entity:
                    return dict(code="success", resource=request['resource'], entity=json.dumps(entity[attr]))
                else:
                    return dict(code="failed", reason="failed to get this resource attribute")
            except:
                logger.exception('failed to read resource')
            return dict(code='failed', reason='failed to get this resource')
        elif params:
            try:
                query, attrs = 'SELECT rid FROM resource WHERE prid=?', [resource]
                if 'like' in params:
                    query += " AND rid LIKE ?"
                    attrs.append(params['like'])
                if 'limit' in params:
                    query += " LIMIT " + params['limit']
                if 'offset' in params:
                    query += " OFFSET " + params['offset']
                if 'order' in params:
                    query += " " + params['order']
                result = self.db.fetchall(query, attrs)
            except:
                logger.exception('failed to read parent resource')
                return dict(code='failed', reason='failed to get child resources')
            response = [(row[0][len(resource)+1:] if row[0].startswith(resource) else row[0]) for row in result]
        else:
            try:
                result = self.db.fetchone('SELECT type, entity FROM resource WHERE rid=?', (resource,))
            except:
                logger.exception('failed to read resource')
                return dict(code='failed', reason='failed to get this resource')
            if result:
                ctype, entity = result[0], json.loads(result[1])
                entity = dict([(k, v) for k, v in entity.iteritems() if not k or k[0] != "_"])
                return dict(code='success', resource=resource, type=ctype, entity=entity)
            try:
                result = self.db.fetchall('SELECT rid FROM resource WHERE prid=?', (resource,))
            except:
                logger.exception('failed to read parent resource')
                return dict(code='failed', reason='failed to get child resources')
            response = [(row[0][len(resource)+1:] if row[0].startswith(resource) else row[0]) for row in result]
        if response:
           return dict(code='success', resource=resource, type='application/json', entity=response)
        return dict(code='failed', reason='no value found for this resource')
    
    def DELETE(self, request):
        resource = request['resource']
        result = self.db.fetchone('SELECT count(rid) FROM resource WHERE prid=?', (resource,))
        if result[0]:
            return dict(code='failed', reason='this parent resource has children')
        self.db.commit('DELETE FROM resource WHERE rid=?', (resource,))
        self.NOTIFY(resource, 'DELETE')
        return dict(code='success')
    
    def SUBSCRIBE(self, request):
        resource = request['resource']
        try:
            self.db.commit('REPLACE INTO subscribe (rid, cid) VALUES (?, ?)', (resource, self.id))
        except:
            logger.exception('failed to replace subscribe')
            return dict(code='failed', reason='failed to subscribe the client to the resource')
        return dict(code='success')
    
    def UNSUBSCRIBE(self, request):
        resource = request['resource']
        try:
            self.db.commit('DELETE FROM subscribe WHERE rid=? AND cid=?', (resource, self.id))
        except:
            logger.exception('failed to delete subscribe')
            return dict(code='failed', reason='failed to unsubscribe the client from the resource')
        return dict(code='success')
    
    # to be overridden by the sub-class if it supports NOTIFY
    def NOTIFY(self, request, method=None):
        pass 
        
    def get_parent(self, resource):
        index = resource.rfind('/')
        return resource[:index] if index >= 0 else ''
        
    def _parse(self, value):
        match = re.match(r'([^\[\?]+)(\[([^\]\?]*)\])?(\?.*)?$', value)
        if not match: return (value, None, None)
        groups = match.groups()
        return (groups[0], groups[2], dict([x.split('=', 1) for x in groups[3][1:].split('&')]) if groups[3] else None)
        
    
